gui/webdisplay/src/TWebWindow.cxx

```diff
@@ -220,6 +365,109 @@ void ROOT::Experimental::TWebWindow::InvokeCallbacks(bool force)
    }
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Add key - procid pair for started window
+/// Key is random number generated when starting new window
+/// procid is special information about starting process which can be used later to halt it
+
+unsigned ROOT::Experimental::TWebWindow::AddProcId(bool batch_mode, const std::string &key, const std::string &procid)
+{
+   std::lock_guard<std::mutex> grd(fConnMutex);
+
+   ++fConnCnt;
+
+   fKeys.emplace_back(std::make_shared<WebConn>(fConnCnt, batch_mode, key, procid));
+
+   return fConnCnt;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Returns true if provided key value already exists (in processes map or in existing connections)
+
+bool ROOT::Experimental::TWebWindow::HasKey(const std::string &key)
+{
+   std::lock_guard<std::mutex> grd(fConnMutex);
+
+   for (auto &&entry : fKeys) {
+      if (entry->fKey == key)
+         return true;
+   }
+
+   for (auto &&conn : fConn) {
+      if (conn->fKey == key)
+         return true;
+   }
+
+   return false;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+/// Check if started process(es) establish connection. After timeout such processed will be killed
+/// Method invoked from http server thread, therefore appropriate mutex must be used on all relevant data
+
+void ROOT::Experimental::TWebWindow::CheckWebKeys()
```

Comment:
`Check` sounds like it might print an error message. Or maybe return `true` if good and `false` if bad. But this function seems to do much more. Maybe `WaitForWebKeysAndHalt()`?

Reply:
Function is not waits - it really regularly check if something went wrong.
It is part of main polling function, which must be regularly called from the appropriate thread.
By default function called every milliseconds